lib/Conversion/TorchToLinalg/IndirectDataMovement.cpp

```diff
@@ -896,6 +896,177 @@ class ConvertAtenUpsampleNearest2dVecOp
 };
 } // namespace
 
+static Value getGradOutputValue(OpBuilder &builder, Location loc,
+                                Value gradOutput, Type gradOutputElemType,
+                                Value numBatch, Value numChannel,
+                                Value inputIndexH, Value inputIndexW,
+                                Value kernelIndexH, Value kernelIndexW,
+                                SmallVector<Value> &gradOutputSizeIndexValues,
+                                SmallVector<Value, 2> &scaleFactorsIntValues) {
+  Value constantOne = builder.create<arith::ConstantIndexOp>(loc, 1);
+
+  Value outputIndexH = builder.create<arith::MulIOp>(
+      loc, inputIndexH, castIntToIndex(builder, loc, scaleFactorsIntValues[0]));
+  outputIndexH = builder.create<arith::AddIOp>(loc, outputIndexH, kernelIndexH);
+
+  Value outputIndexW = builder.create<arith::MulIOp>(
+      loc, inputIndexW, castIntToIndex(builder, loc, scaleFactorsIntValues[1]));
+  outputIndexW = builder.create<arith::AddIOp>(loc, outputIndexW, kernelIndexW);
+
+  // Handling corner cases.
+  Value gradOutputHMinusOne = builder.create<arith::SubIOp>(
+      loc, gradOutputSizeIndexValues[2], constantOne);
+  Value predH = builder.create<arith::CmpIOp>(
+      loc, arith::CmpIPredicate::sle, outputIndexH, gradOutputHMinusOne);
+  outputIndexH = builder.create<arith::SelectOp>(loc, predH, outputIndexH,
+                                                 gradOutputHMinusOne);
+
+  Value gradOutputWMinusOne = builder.create<arith::SubIOp>(
+      loc, gradOutputSizeIndexValues[3], constantOne);
+  Value predW = builder.create<arith::CmpIOp>(
+      loc, arith::CmpIPredicate::sle, outputIndexW, gradOutputWMinusOne);
+  outputIndexW = builder.create<arith::SelectOp>(loc, predW, outputIndexW,
+                                                 gradOutputWMinusOne);
+
+  Value gradOutputValue = builder.create<tensor::ExtractOp>(
+      loc, gradOutput,
+      ValueRange{numBatch, numChannel, outputIndexH, outputIndexW});
+  Value constantZero =
+      builder.create<arith::ConstantOp>(loc, builder.getF32FloatAttr(0.0));
+  Value pred = builder.create<arith::AndIOp>(loc, predH, predW);
+  Value result = builder.create<arith::SelectOp>(
+      loc, pred, gradOutputValue,
+      convertScalarToDtype(builder, loc, constantZero, gradOutputElemType));
+
+  return result;
+}
+
+namespace {
+class ConvertAtenUpsampleNearest2dBackwardVecOp
```

Comment:
Can you add the python pseudo code you have in the PR description as a comment here? I find it makes understanding the code much easier.